src/MongoDB.Driver/WriteModel.cs

```diff
@@ -77,20 +77,22 @@ private static WriteModel<TDocument> ConvertInsertRequest(InsertRequest request)
 
         private static WriteModel<TDocument> ConvertUpdateRequest(UpdateRequest request)
         {
+            var updateDocument = GetUpdateDocument(request.Update);
```

Comment:
According to comments on line 33 at the beginning of this file, the static methods from this file are used only in `Legacy` code. Since we won't support Legacy Api in the scope of this ticket, I removed most of the changes in this file. Currently, we expect only `BsonDocument` in `Update`. Otherwise, we will have `NotSupportedException` error

Reply:
I don't think those comments are correct.

These conversions from a Core `UpdateRequest` to a High `WriteModel` are NOT used by Legacy. In fact the `WriteModel` class is in the new High level.

Reply:
I think this method needs to support pipeline updates also. It only needs to be slightly modified to do so (with most of the changes in `UnwrapUpdate` instead):

    private static WriteModel<TDocument> ConvertUpdateRequest(UpdateRequest request)
    {
        if (request.IsMulti)
        {
            return new UpdateManyModel<TDocument>(UnwrapFilter(request.Filter), UnwrapUpdate(request.Update))
            {
                ArrayFilters = request.ArrayFilters == null ? null : new List<ArrayFilterDefinition>(request.ArrayFilters.Select(f => new BsonDocumentArrayFilterDefinition<BsonValue>(f))),
                Collation = request.Collation,
                IsUpsert = request.IsUpsert
            };
        }

        if (!IsReplaceOneUpdate(request.Update))
        {
            return new UpdateOneModel<TDocument>(UnwrapFilter(request.Filter), UnwrapUpdate(request.Update))
            {
                ArrayFilters = request.ArrayFilters == null ? null : new List<ArrayFilterDefinition>(request.ArrayFilters.Select(f => new BsonDocumentArrayFilterDefinition<BsonValue>(f))),
                Collation = request.Collation,
                IsUpsert = request.IsUpsert
            };
        }

        return ConvertToReplaceOne(request);
    }

Reply:
New helper method:

    private static bool IsReplaceOneUpdate(BsonValue update)
    {
        if (update is BsonDocument updateDocument)
        {
            var firstElementName = updateDocument.GetElement(0).Name;
            return !firstElementName.StartsWith("$");
        }

        if (update is BsonArray)
        {
            return false;
        }

        return true;
    }

Reply:
I can't comment on UnwrapUpdate below because it didn't change, so I'll comment here that it can be refactored to support pipelines like this:

    private static UpdateDefinition<TDocument> UnwrapUpdate(BsonValue update)
    {
        var wrapper = update as BsonDocumentWrapper;
        if (wrapper != null)
        {
            if (wrapper.Wrapped is BsonDocument)
            {
                return new BsonDocumentUpdateDefinition<TDocument>((BsonDocument)wrapper.Wrapped);
            }
            return new ObjectUpdateDefinition<TDocument>(wrapper.Wrapped);
        }

        if (update is BsonArray stages)
        {
            var pipeline = new BsonDocumentStagePipelineDefinition<TDocument, TDocument>(stages.Cast<BsonDocument>());
            return new PipelineUpdateDefinition<TDocument>(pipeline);
        }

        return new BsonDocumentUpdateDefinition<TDocument>((BsonDocument)update);
    }